### ISSUE TITLE
Predicting sub classes

### ISSUE THREAD

**Issue by CodeJjang** [jwyang/faster-rcnn.pytorch#104]
I'd like to be able to predict not only regular classes (i.e. car, ship, etc) but also subclasses (color, type of car, type of ship, and more types of features).  
Do you have an idea how can I embed it into this model?  
Note that not all the classes have subclasses, only a subset. Also, classes might have different subclasses (i.e. cars would have a list of car types, a color, and several car features, ships would have a list of ship types only, etc).

I guess the behavior should be: First predict a regular class, then, dependeing on the prediction class (car, ship...) further predict subclasses, but I can't think of how to add this behavior to this model.

One option that I think of, is to add another several feed forward networks in the end, each responsible for predicting a type of subclass labels. For example, a network for predicting color, a network for predicting car types, a network for predicting car features, a network for predicting ship types, etc. Then, when the network predicts a class which have subclasses, the matching feed forward networks (let's say, for a car, there would be 3) will be further trained and their loss would be kept and tracked independently.  
I don't know however if there's an easier solution or if it's even going to work. Would love to hear if you have other ideas, and what you think about that one as well.

**Comment by jwyang** [comment]
Hi, @CodeJjang it is more like an attribute recognition system. So you can plug into some sub-networks to predict the attributes given the pooled feature. I think your proposed way would work. you can have a try.

**Comment by CodeJjang** [comment]
Thanks

**Comment by CodeJjang** [comment]
@jwyang I started implementing this but I tackle many issues.  
I started with adding another class type  of color, which as I said exists for only two of my main detection clases, called Coarse Grained classes (large and small cars, for example).  

I had to add the GT color label in all the places together with the Coarse Grained classes, in order to have correlation between the trainining example and the appropriate color.  
This starts in the `roidb`, where you hold a tensor of `<GT_label, BB_x1, BB_y1m....>` and continues in the `minibatch` where you hold a similar tensor, so their length is now 6 instead of 5 (GT Coarse Grained Label, GT Color Label, BB_x1,....).
I added, of course, another FF layer just like the FF classification layer in the ResNet model.

But unfortunately, in many places in the code you relied on the fact that those tensors are of size 5 (gt label + 4 bbox points), and also, that indices 1:5 consists of the BB points,  so I had to change this in many places.

However despite all the effort, the model now fails to learn also the Coarse Grained labels, which it did succeed before the changing.  
I think there are some inner calculations I missed (maybe in the C code?) that relied on the original tensor structure that I changed so now they calculate wrong results (I suspect on the `ROI alignment`).  

I'd like to hear your opinion on how to easily add the extra class (and then after I add 1 class, I'd be able to add several more), with **minimum** altering to your code, in order to prevent such bugs.  
I don't see a way to do this without touching the `roidb` and `minibatch` structures, which then detroys many calculations.
